File: 0042-trapping-rain-water/0042-trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */

// Approach 2: Two pointers
// TC: O(n), SP: O(1)
var trap = function (height) {
    const n = height.length;

    if (n < 3) return 0;

    let left = 0, right = n - 1;
    let result = 0, maxLeft = 0, maxRight = 0;

    while (left <= right) {
        if (height[left] <= height[right]) {
            if (height[left] >= maxLeft) {
                maxLeft = height[left];
            } else {
                result += maxLeft - height[left];
            }
            left++;
        } else {
            if (height[right] >= maxRight) {
                maxRight = height[right];
            } else {
                result += maxRight - height[right];
            }
            right--;
        }
    }

    console.log(`Min: ${maxLeft}, right: ${maxRight}`);
    return result;
}



// Approach 1: Brute force
// TC: O(n ^ 2): The outer loop runs n − 2, and each inner loop runs up to \U0001d45b n times.
// SP: O(1)
var trap1 = function (height) {
    const n = height.length;

    if (n < 3) return 0;

    let result = 0;

    //Skipped calculations for the first and last bars since they cannot trap water.
    for (let i = 1; i < n - 1; i++) {
        let maxLeft = 0, maxRight = 0;

        // Calculate maxLeft for the current position
        for (let j = 0; j < i; j++) {
            maxLeft = Math.max(maxLeft, height[j]);
        }

        // Calculate maxRight for the current position
        for (let j = i; j < n; j++) {
            maxRight = Math.max(maxRight, height[j]);
        }

        // Calculate water trapped at the current position
        const water = Math.min(maxLeft, maxRight) - height[i];
        if (water > 0)
            result += water;
    }

    return result;
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    
    def trap(self, height: List[int]) -> int:
        n = len(height)
        left_max = [0] * n
        right_max = [0] * n

        left_max[0] = height[0]
        for i in range(1, n):
            left_max[i] = max(left_max[i - 1], height[i]) 
        
        right_max[n - 1] = height[n - 1]
        for i in range(n - 2, -1, -1):
            right_max[i] = max(right_max[i + 1], height[i])
        
        water = 0
        for i in range(n):
            water += min(left_max[i], right_max[i]) - height[i]
        
        return water



    # Brute - T: O(N ^ 2), Space: O(1)
    def trap1(self, height: List[int]) -> int:
        n = len(height)

        water = 0

        for i in range(n):
            left = max(height[:i + 1])
            right = max(height[i: ])

            water += min(left, right) - height[i]
        
        return water
        